File: frontend/src/types/visionTest.ts
// Types for the Vision Profile Test Module (simple 10-question version).
// These mirror backend/app/schemas/vision_test.py field-for-field — keep
// them in sync manually (see docs/roadmap.md for a future codegen step).

/** The four broad categories every question tests perception of. */
export type ColorCategory = 'red' | 'green' | 'blue' | 'yellow' | 'orange' | 'purple' | 'brown' | 'gray'

export type QuestionDifficulty = 'basic' | 'intermediate' | 'real_world'

export type QuestionAxis = 'red_green' | 'blue_yellow' | 'mixed' | 'real_world'

/** One answer option shown as a button (color name + swatch). */
export interface AnswerOption {
  id: string
  label: string
  hex: string
}

/** One question in the 10-question battery. */
export interface VisionTestQuestion {
  id: string
  index: number // 1-10, display order
  axis: QuestionAxis
  difficulty: QuestionDifficulty
  prompt: string
  /** The color patch shown to the user. */
  stimulus_hex: string
  /** 4-5 answer choices, in display order. Exactly one is correct. */
  options: AnswerOption[]
  /** Which option.id is correct — sent to the client so scoring CAN happen
   *  client-side for instant per-question feedback if desired, but the
   *  authoritative vision-profile scoring always happens server-side in
   *  /api/vision-test/submit. Never trust only the client's own scoring. */
  correct_option_id: string
}

export interface VisionTestConfig {
  test_id: string
  version: string
  time_limit_seconds: number // 120 = 2 minutes
  questions: VisionTestQuestion[]
}

/** One submitted answer for one question. */
export interface UserAnswer {
  question_id: string
  selected_option_id: string
  response_time_ms: number
}

export interface SubmitVisionTestRequest {
  user_id: string
  test_id: string
  answers: UserAnswer[]
}

export type DeficiencyType = 'red-green' | 'blue-yellow' | 'none' | 'unknown'
export type Severity = 'none' | 'mild' | 'moderate' | 'severe'

export interface PerceptionScores {
  red: number
  green: number
  blue: number
  yellow: number
}

export interface ColorTransformation {
  from: string
  to: string
  reason: string
}

/** The Vision Map — exact shape requested in the spec. */
export interface VisionProfile {
  user_id: string
  deficiency_type: DeficiencyType
  severity: Severity
  perception_scores: PerceptionScores
  recommended_transformations: ColorTransformation[]
  created_at: string
  updated_at: string
}

export interface SubmitVisionTestResponse {
  profile: VisionProfile
  score_summary: {
    total_questions: number
    correct_count: number
    accuracy: number
  }
}

File: backend/app/schemas/vision_test.py
"""
Schemas for the Vision Profile Test Module (simple 10-question version).
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


# --- /api/vision-test/start -------------------------------------------------

class AnswerOptionOut(BaseModel):
    id: str
    label: str
    hex: str


class QuestionOut(BaseModel):
    id: str
    index: int
    axis: str
    difficulty: str
    prompt: str
    stimulus_hex: str
    options: list[AnswerOptionOut]
    correct_option_id: str


class VisionTestConfigOut(BaseModel):
    test_id: str
    version: str
    time_limit_seconds: int
    questions: list[QuestionOut]


# --- /api/vision-test/submit ------------------------------------------------

class UserAnswerIn(BaseModel):
    question_id: str
    selected_option_id: str
    response_time_ms: int = Field(ge=0, le=120_000)


class SubmitVisionTestIn(BaseModel):
    user_id: str
    test_id: str
    answers: list[UserAnswerIn]


class PerceptionScoresOut(BaseModel):
    red: float
    green: float
    blue: float
    yellow: float
    brown: float | None = 1.0


class ColorTransformationOut(BaseModel):
    from_: str = Field(alias="from")
    to: str
    reason: str

    model_config = ConfigDict(populate_by_name=True)


class MeaningBasedTransformationOut(BaseModel):
    original_color_name: str
    transformed_color_hex: str
    meaning_label: str
    explanation: str


class PersonalImpactOut(BaseModel):
    workplace: str
    productivity: str
    dashboard: str
    daily: str


class VisionProfileOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: str
    deficiency_type: str
    deficiency_name: str | None = None
    clinical_diagnosis: str | None = None
    severity: str
    percent_accuracy: int | None = None
    color_confusion_status: str | None = None
    perception_scores: PerceptionScoresOut
    recommended_transformations: list[ColorTransformationOut] = []
    meaning_based_transformations: list[MeaningBasedTransformationOut] = []
    risk_areas: list[str] = []
    personal_impact: PersonalImpactOut | None = None
    ai_explanation: str | None = None
    created_at: datetime
    updated_at: datetime


class ScoreSummaryOut(BaseModel):
    total_questions: int
    correct_count: int
    accuracy: float


class SubmitVisionTestOut(BaseModel):
    profile: VisionProfileOut
    score_summary: ScoreSummaryOut
